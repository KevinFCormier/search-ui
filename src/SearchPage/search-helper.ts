export function formatSearchbarSuggestions(data: string[], suggestionKind: 'label' | 'filter' | 'value') {
    return data.map(field => {
        return {
          id: `id-${field}`,
          key:`key-${field}`,
          name: field,
          kind: suggestionKind
        }
      })
}

export const convertStringToQuery = (searchText: string) => {
    const searchTokens = searchText.split(' ')
    const keywords = searchTokens.filter(token => token !== '' && token.indexOf(':') < 0)
    const filters = searchTokens.filter(token => token.indexOf(':') >= 0)
      .map(f => {
        const splitIdx = f.indexOf(':')
        const property = f.substring(0, splitIdx)
        const values = f.substring(splitIdx + 1)
        return { property, values: values.split(',') }
      })
      .filter(f => ['', '=', '<', '>', '<=', '>=', '!=', '!'].findIndex(op => op === f.values[0]) === -1)
    return {
      keywords,
      filters,
      limit: 10000 // TODO make configurable config (config['searchQueryLimit'])
    }
  }

  export const formatNumber = (count: number) => {
    if (count > 999) {
      // show one decimal place
      const num = (count - (count % 100)) / 1000
      return num + 'k'
    }
    return count
  }