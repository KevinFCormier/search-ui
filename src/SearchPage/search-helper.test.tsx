import React from 'react'
import { render } from '@testing-library/react'
import { formatSearchbarSuggestions, convertStringToQuery, formatNumber } from './search-helper'

test('Correctly returns formatSearchbarSuggestions', () => {
    const testData = [
        'kind',
        'cluster',
        'deployment'
    ]
    const result = formatSearchbarSuggestions(testData, 'filter')
    expect(result).toMatchSnapshot()
})

test('Correctly returns convertStringToQuery', () => {
    const testData = 'namespace:open-cluster-management kind:pod'
    const result = convertStringToQuery(testData)
    expect(result).toMatchSnapshot()
})

test('Correctly returns formatNumber of number larger than 999', () => {
    const testData = 1000
    const result = formatNumber(testData)
    expect(result).toMatchSnapshot()
})
test('Correctly returns formatNumber of number smaller than 999', () => {
    const testData = 10
    const result = formatNumber(testData)
    expect(result).toMatchSnapshot()
})
