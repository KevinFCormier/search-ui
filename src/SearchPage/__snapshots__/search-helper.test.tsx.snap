// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly returns convertStringToQuery 1`] = `
Object {
  "filters": Array [
    Object {
      "property": "namespace",
      "values": Array [
        "open-cluster-management",
      ],
    },
    Object {
      "property": "kind",
      "values": Array [
        "pod",
      ],
    },
  ],
  "keywords": Array [],
}
`;

exports[`Correctly returns formatSearchbarSuggestions 1`] = `
Array [
  Object {
    "id": "id-kind",
    "key": "key-kind",
    "kind": "filter",
    "name": "kind",
  },
  Object {
    "id": "id-cluster",
    "key": "key-cluster",
    "kind": "filter",
    "name": "cluster",
  },
  Object {
    "id": "id-deployment",
    "key": "key-deployment",
    "kind": "filter",
    "name": "deployment",
  },
]
`;
